File: rlm/cli/service.py
"""RLM Service Manager facade.

Mantém a API pública estável do CLI, mas delega a implementação para
submódulos menores de runtime, installers e update.
"""

from __future__ import annotations

import json
import os
import sys
from pathlib import Path
from typing import Any, Callable, cast

from rlm.cli.context import CliContext
from rlm.cli.json_output import build_cli_json_envelope
from rlm.cli.output import HAS_RICH, _err, _info, _ok, _warn
from rlm.cli.state.diagnosis import build_launcher_state_diagnosis
from rlm.cli.state.launcher import (
    mark_bootstrap_success,
    mark_daemon_installed,
    mark_runtime_status,
    mark_stopped,
    mark_update,
    summarize_launcher_state,
)
from rlm.cli.state.pid import pid_alive as _pid_alive, read_pid_file as _read_pid, remove_pid as _remove_pid, write_pid as _write_pid
from rlm.cli.service_installers import install_launchd_service_impl, install_systemd_service_impl
from rlm.cli.service_runtime import (
    ServiceRuntimeLayout,
    port_accepting_connections,
    runtime_mode,
    services_are_running as services_are_running_impl,
    show_runtime_status,
    start_runtime,
    stop_runtime,
)
from rlm.cli.service_update import update_installation_impl
from rlm.cli.service_wireguard import add_wireguard_peer_impl


# --------------------------------------------------------------------------- #
# Configuração  (lazy — evita side-effect no import)                           #
# --------------------------------------------------------------------------- #

_SERVICE_NAME = "rlm"

# Patchable overrides — testes usam patch("rlm.cli.service._PID_DIR", ...) etc.
_PID_DIR: Path | None = None
_PID_API: Path | None = None
_PID_WS: Path | None = None
_LOG_DIR: Path | None = None


def _default_paths() -> tuple[Path, Path]:
    """Retorna (pid_dir, log_dir) do contexto padrão sem load_env."""
    ctx = CliContext.from_environment(load_env=False)
    return ctx.paths.runtime_dir, ctx.paths.log_dir


def _runtime_layout() -> ServiceRuntimeLayout:
    if _PID_DIR is not None:
        pid_dir = _PID_DIR
        log_dir = _LOG_DIR or pid_dir.parent / "log"
    else:
        pid_dir, log_dir = _default_paths()
    return ServiceRuntimeLayout(
        pid_dir=pid_dir,
        pid_api=_PID_API or pid_dir / "api.pid",
        pid_ws=_PID_WS or pid_dir / "ws.pid",
        log_dir=_LOG_DIR or log_dir,
        service_name=_SERVICE_NAME,
    )


# --------------------------------------------------------------------------- #
# Rich Table (re-export para show_status)                                      #
# --------------------------------------------------------------------------- #

try:
    from rich.table import Table
except ImportError:
    Table = None


def _load_env_settings() -> dict[str, str]:
    context = CliContext.from_environment(load_env=True)
    return dict(context.env)


def _service_context(context: CliContext | None = None, *, load_env: bool = True) -> CliContext:
    if context is not None:
        if load_env:
            context.load_env_file(override=False)
        return context
    return CliContext.from_environment(load_env=load_env)


def _os_getuid() -> int:
    getter = getattr(os, "getuid", None)
    if callable(getter):
        return cast(Callable[[], int], getter)()
    return 0


def _os_geteuid() -> int:
    getter = getattr(os, "geteuid", None)
    if callable(getter):
        return cast(Callable[[], int], getter)()
    return 0


def start_services(
    foreground: bool = False,
    api_only: bool = False,
    ws_only: bool = False,
    context: CliContext | None = None,
) -> int:
    current_context = _service_context(context, load_env=True)
    rc = start_runtime(
        current_context,
        _runtime_layout(),
        foreground=foreground,
        api_only=api_only,
        ws_only=ws_only,
        write_pid=_write_pid,
        remove_pid=_remove_pid,
        ok=_ok,
        warn=_warn,
        err=_err,
        info=_info,
    )
    if rc == 0:
        mark_bootstrap_success(
            current_context,
            source="start",
            mode=runtime_mode(foreground=foreground, api_only=api_only, ws_only=ws_only),
            api_enabled=not ws_only,
            ws_enabled=ws_only or not api_only,
        )
    return rc


def stop_services(context: CliContext | None = None) -> int:
    current_context = _service_context(context, load_env=False)
    rc = stop_runtime(
        _runtime_layout(),
        read_pid=_read_pid,
        remove_pid=_remove_pid,
        pid_alive=_pid_alive,
        ok=_ok,
        warn=_warn,
        err=_err,
        info=_info,
    )
    if rc == 0:
        mark_stopped(current_context)
    return rc


def _build_status_snapshot(context: CliContext) -> dict[str, Any]:
    layout = _runtime_layout()
    api_pid = _read_pid(layout.pid_api)
    ws_pid = _read_pid(layout.pid_ws)
    api_running = bool(api_pid and _pid_alive(api_pid))
    ws_running = bool(ws_pid and _pid_alive(ws_pid))
    api_port_open = port_accepting_connections(context.api_host(), context.api_port())
    ws_port_open = port_accepting_connections(context.ws_host(), context.ws_port())
    launcher = build_launcher_state_diagnosis(context, health_online=api_port_open)
    state = mark_runtime_status(context, api_running=api_running, ws_running=ws_running)

    return {
        "runtime": {
            "api": {
                "pid": api_pid,
                "running": api_running,
                "port_open": api_port_open,
                "url": f"{context.api_base_url()}/",
                "docs_url": context.docs_url(),
                "log_file": str((layout.log_dir / "api.log").resolve()),
            },
            "ws": {
                "pid": ws_pid,
                "running": ws_running,
                "port_open": ws_port_open,
                "url": context.ws_base_url(),
                "log_file": str((layout.log_dir / "ws.log").resolve()),
            },
            "webchat_url": context.webchat_url(),
        },
        "launcher_state": launcher,
        "persisted_state_summary": summarize_launcher_state(state),
    }


def show_status(context: CliContext | None = None, json_output: bool = False) -> int:
    current_context = _service_context(context, load_env=True)
    if json_output:
        payload = _build_status_snapshot(current_context)
        envelope = build_cli_json_envelope(
            "status",
            payload,
            severity=str(payload.get("launcher_state", {}).get("severity", "info")),
        )
        print(json.dumps(envelope, indent=2, ensure_ascii=False))
        return 0

    rc, api_running, ws_running = show_runtime_status(
        current_context,
        _runtime_layout(),
        read_pid=_read_pid,
        pid_alive=_pid_alive,
        info=_info,
        warn=_warn,
        has_rich=HAS_RICH,
        table_cls=Table,
    )
    state = mark_runtime_status(current_context, api_running=api_running, ws_running=ws_running)
    _info(f"Launcher state: {summarize_launcher_state(state)}")
    return rc


def _services_are_running() -> bool:
    return services_are_running_impl(_runtime_layout(), read_pid=_read_pid, pid_alive=_pid_alive)


def update_installation(
    check_only: bool = False,
    restart: bool = True,
    target_path: str | None = None,
    context: CliContext | None = None,
) -> int:
    current_context = _service_context(context, load_env=True)
    rc = update_installation_impl(
        current_context,
        check_only=check_only,
        restart=restart,
        target_path=target_path,
        info=_info,
        ok=_ok,
        err=_err,
        services_are_running=_services_are_running,
        stop_services=lambda: stop_services(context=current_context),
        start_services=lambda: start_services(foreground=False, context=current_context),
    )
    if rc == 0 and not check_only:
        mark_update(current_context, restarted=restart and _services_are_running())
    return rc


def install_systemd_service(project_root: Path, env_path: Path) -> int:
    rc, unit_file = install_systemd_service_impl(
        project_root,
        env_path,
        service_name=_SERVICE_NAME,
        ok=_ok,
        err=_err,
        info=_info,
    )
    if rc == 0:
        current_context = CliContext.from_environment(load_env=False)
        mark_daemon_installed(
            current_context,
            manager="systemd",
            definition_path=unit_file,
            project_root=project_root,
            env_path=env_path,
        )
    return rc


def install_launchd_service(project_root: Path, env_path: Path) -> int:
    _, log_dir = _default_paths()
    rc, plist_path = install_launchd_service_impl(
        project_root,
        env_path,
        log_dir,
        os_getuid=_os_getuid,
        ok=_ok,
        info=_info,
    )
    if rc == 0:
        current_context = CliContext.from_environment(load_env=False)
        mark_daemon_installed(
            current_context,
            manager="launchd",
            definition_path=plist_path,
            project_root=project_root,
            env_path=env_path,
        )
    return rc


# --------------------------------------------------------------------------- #
# WireGuard peer add                                                           #
# --------------------------------------------------------------------------- #

def add_wireguard_peer(name: str, pubkey: str, ip: str) -> int:
    return add_wireguard_peer_impl(
        name,
        pubkey,
        ip,
        wg_conf=Path("/etc/wireguard/wg0.conf"),
        os_geteuid=_os_geteuid,
        ok=_ok,
        warn=_warn,
        err=_err,
        info=_info,
    )

File: packages/cli/src/service.ts
/**
 * service.ts — Facade do Service Manager do RLM.
 *
 * Porta fiel de rlm/cli/service.py
 * Delega para service-runtime, service-installers, service-update e service-wireguard.
 */

import { join } from "node:path";
import { c } from "./format.js";
import { buildCliJsonEnvelope } from "./json-output.js";
import { CliContext } from "./context.js";
import {
  buildRuntimeLayout,
  ServiceRuntimeLayout,
  runtimeMode,
  portAcceptingConnections,
  startRuntime,
  stopRuntime,
  showRuntimeStatus,
  servicesAreRunning,
} from "./service-runtime.js";
import {
  installSystemdService,
  installLaunchdService,
} from "./service-installers.js";
import { updateInstallation } from "./service-update.js";
import { addWireguardPeer as _addWireguardPeer } from "./service-wireguard.js";
import {
  markBootstrapSuccess,
  markDaemonInstalled,
  markRuntimeStatus,
  markStopped,
  markUpdate,
  summarizeLauncherState,
} from "./state/launcher.js";
import { buildLauncherStateDiagnosis } from "./state/diagnosis.js";

const SERVICE_NAME = "rlm";

// ---------------------------------------------------------------------------
// callbacks padrão (chalk)
// ---------------------------------------------------------------------------

const callbacks = {
  ok: (msg: string) => console.log(c.success(msg)),
  warn: (msg: string) => console.warn(c.warn(msg)),
  err: (msg: string) => console.error(c.error(msg)),
  info: (msg: string) => console.log(c.info(msg)),
};

// ---------------------------------------------------------------------------
// helpers internos
// ---------------------------------------------------------------------------

function getContext(ctx?: CliContext, loadEnv = true): CliContext {
  if (ctx) {
    if (loadEnv) ctx.loadEnvFile({ override: false });
    return ctx;
  }
  return CliContext.fromEnvironment({ loadEnv });
}

function getRuntimeLayout(ctx?: CliContext): ServiceRuntimeLayout {
  const context = ctx ?? CliContext.fromEnvironment({ loadEnv: false });
  return buildRuntimeLayout(context.paths.stateRoot, SERVICE_NAME);
}

// ---------------------------------------------------------------------------
// startServices
// ---------------------------------------------------------------------------

export async function startServices(opts: {
  foreground?: boolean;
  apiOnly?: boolean;
  wsOnly?: boolean;
  context?: CliContext;
} = {}): Promise<number> {
  const { foreground = false, apiOnly = false, wsOnly = false, context } = opts;
  const ctx = getContext(context, true);
  const layout = getRuntimeLayout(ctx);

  const rc = await startRuntime(ctx, layout, { foreground, apiOnly, wsOnly }, callbacks);
  if (rc === 0) {
    markBootstrapSuccess(ctx, {
      source: "start",
      mode: runtimeMode({ foreground, apiOnly, wsOnly }),
      apiEnabled: !wsOnly,
      wsEnabled: wsOnly || !apiOnly,
    });
  }
  return rc;
}

// ---------------------------------------------------------------------------
// stopServices
// ---------------------------------------------------------------------------

export async function stopServices(opts: { context?: CliContext } = {}): Promise<number> {
  const ctx = getContext(opts.context, false);
  const layout = getRuntimeLayout(ctx);
  const rc = await stopRuntime(layout, callbacks);
  if (rc === 0) markStopped(ctx);
  return rc;
}

// ---------------------------------------------------------------------------
// showStatus
// ---------------------------------------------------------------------------

export async function showStatus(opts: {
  context?: CliContext;
  jsonOutput?: boolean;
} = {}): Promise<number> {
  const { jsonOutput = false, context } = opts;
  const ctx = getContext(context, true);
  const layout = getRuntimeLayout(ctx);

  if (jsonOutput) {
    const apiPortOpen = await portAcceptingConnections(ctx.apiHost(), ctx.apiPort());
    const diagnosis = await buildLauncherStateDiagnosis(ctx, { healthOnline: apiPortOpen });

    const { readPidFile, pidAlive } = await import("./state/pid.js");
    const apiPid = readPidFile(layout.pidApi);
    const wsPid = readPidFile(layout.pidWs);
    const apiRunning = apiPid !== null && pidAlive(apiPid);
    const wsRunning = wsPid !== null && pidAlive(wsPid);
    const wsPortOpen = await portAcceptingConnections(ctx.wsHost(), ctx.wsPort());

    const state = markRuntimeStatus(ctx, { apiRunning, wsRunning });
    const payload = {
      runtime: {
        api: {
          pid: apiPid,
          running: apiRunning,
          portOpen: apiPortOpen,
          url: `${ctx.apiBaseUrl()}/`,
          docsUrl: ctx.docsUrl(),
          logFile: layout.apiLog,
        },
        ws: {
          pid: wsPid,
          running: wsRunning,
          portOpen: wsPortOpen,
          url: ctx.wsBaseUrl(),
          logFile: layout.wsLog,
        },
        webchatUrl: ctx.webchatUrl(),
      },
      launcherState: diagnosis,
      persistedStateSummary: summarizeLauncherState(state),
    };

    const envelope = buildCliJsonEnvelope("status", payload, diagnosis.severity === "warning" ? "warn" : "info");
    process.stdout.write(JSON.stringify(envelope, null, 2) + "\n");
    return 0;
  }

  const { exitCode, apiRunning, wsRunning } = await showRuntimeStatus(ctx, layout, callbacks);
  const state = markRuntimeStatus(ctx, { apiRunning, wsRunning });
  callbacks.info(`Launcher state: ${summarizeLauncherState(state)}`);
  return exitCode;
}

// ---------------------------------------------------------------------------
// installDaemon
// ---------------------------------------------------------------------------

export function installDaemon(opts: { context?: CliContext } = {}): number {
  const ctx = getContext(opts.context, true);
  const layout = getRuntimeLayout(ctx);
  const { projectRoot, envPath, logDir } = {
    projectRoot: ctx.paths.projectRoot,
    envPath: ctx.paths.envPath,
    logDir: layout.logDir,
  };

  if (process.platform === "linux") {
    const result = installSystemdService({
      projectRoot,
      envPath,
      logDir,
      serviceName: SERVICE_NAME,
      callbacks,
    });
    if (result.exitCode === 0) {
      markDaemonInstalled(ctx, {
        manager: "systemd",
        definitionPath: result.unitPath,
        projectRoot,
        envPath,
      });
    }
    return result.exitCode;
  }

  if (process.platform === "darwin") {
    const result = installLaunchdService({
      projectRoot,
      envPath,
      logDir,
      callbacks,
    });
    if (result.exitCode === 0) {
      markDaemonInstalled(ctx, {
        manager: "launchd",
        definitionPath: result.unitPath,
        projectRoot,
        envPath,
      });
    }
    return result.exitCode;
  }

  callbacks.err(`Instalação automática de daemon não suportada em ${process.platform}`);
  return 1;
}

// ---------------------------------------------------------------------------
// updateInstallationFacade
// ---------------------------------------------------------------------------

export async function updateInstallationFacade(opts: {
  checkOnly?: boolean;
  restart?: boolean;
  targetPath?: string;
  context?: CliContext;
} = {}): Promise<number> {
  const { checkOnly = false, restart = true, targetPath, context } = opts;
  const ctx = getContext(context, true);
  const layout = getRuntimeLayout(ctx);

  const rc = await updateInstallation(
    ctx,
    { checkOnly, restart, targetPath },
    callbacks,
    {
      servicesAreRunning: () => servicesAreRunning(layout),
      stopServices: () => stopServices({ context: ctx }),
      startServices: () => startServices({ context: ctx }),
    },
  );
  if (rc === 0 && !checkOnly) {
    markUpdate(ctx, { restarted: restart && servicesAreRunning(layout) });
  }
  return rc;
}

// ---------------------------------------------------------------------------
// addWireguardPeer
// ---------------------------------------------------------------------------

export function addWireguardPeerFacade(name: string, pubkey: string, ip: string): number {
  return _addWireguardPeer(name, pubkey, ip, "/etc/wireguard/wg0.conf", callbacks);
}
